broker/src/main/java/io/camunda/zeebe/broker/bootstrap/CommandApiServiceStep.java

```diff
@@ -0,0 +1,202 @@
+              startCommandApiService(brokerStartupContext, startupFuture);
+            },
+            startupFuture));
+  }
+
+  private void startCommandApiService(
+      final BrokerStartupContext brokerStartupContext,
+      final ActorFuture<BrokerStartupContext> startupFuture) {
+
+    final var concurrencyControl = brokerStartupContext.getConcurrencyControl();
+    final var brokerInfo = brokerStartupContext.getBrokerInfo();
+    final var brokerCfg = brokerStartupContext.getBrokerConfig();
+    final var schedulingService = brokerStartupContext.getActorSchedulingService();
+    final var messagingService = brokerStartupContext.getCommandApiMessagingService();
+
+    final var transportFactory = new TransportFactory(schedulingService);
+
+    final var serverTransport =
+        transportFactory.createServerTransport(brokerInfo.getNodeId(), messagingService);
+
+    final BackpressureCfg backpressureCfg = brokerCfg.getBackpressure();
+    PartitionAwareRequestLimiter limiter = PartitionAwareRequestLimiter.newNoopLimiter();
+    if (backpressureCfg.isEnabled()) {
+      limiter = PartitionAwareRequestLimiter.newLimiter(backpressureCfg);
+    }
+
+    final var commandApiService = new CommandApiServiceImpl(serverTransport, brokerInfo, limiter);
+
+    concurrencyControl.runOnCompletion(
+        schedulingService.submitActor(commandApiService),
+        proceed(
+            () -> {
+              brokerStartupContext.setCommandApiService(commandApiService);
+              brokerStartupContext.addPartitionListener(commandApiService);
+              brokerStartupContext.addDiskSpaceUsageListener(commandApiService);
+              startupFuture.complete(brokerStartupContext);
+            },
+            startupFuture));
+  }
+
+  private void closeServerTransport(
+      final BrokerStartupContext brokerShutdownContext,
+      final ConcurrencyControl concurrencyControl,
+      final ActorFuture<BrokerStartupContext> shutdownFuture) {
+    final var serverTransport = brokerShutdownContext.getCommandApiServerTransport();
+
+    if (serverTransport == null) {
+      stopMessagingService(brokerShutdownContext, shutdownFuture);
+      return;
+    }
+
+    concurrencyControl.runOnCompletion(
+        serverTransport.closeAsync(),
+        proceed(
+            () -> {
+              brokerShutdownContext.setCommandApiServerTransport(null);
+              stopMessagingService(brokerShutdownContext, shutdownFuture);
+            },
+            shutdownFuture));
+  }
+
+  private void stopMessagingService(
+      final BrokerStartupContext brokerShutdownContext,
+      final ActorFuture<BrokerStartupContext> shutdownFuture) {
+
+    final var messagingService = brokerShutdownContext.getCommandApiMessagingService();
+
+    if (messagingService == null) {
+      shutdownFuture.complete(brokerShutdownContext);
+      return;
+    }
+
+    messagingService
+        .stop()
+        .whenComplete(
```

Comment:
:x: Same as above, potentially a race condition as the context is mutated by multiple threads.